pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go

```diff
@@ -18,70 +18,191 @@ package k8sorchestrator
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+
 	"sigs.k8s.io/vsphere-csi-driver/pkg/common/config"
 	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
 	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
 )
 
 var (
-	k8sOrchestratorInstance         *K8sOrchestrator
-	onceFork8sOrchestratorInstance  sync.Once
-	featureStatesConfigMapName      string
-	featureStatesConfigMapNamespace string
+	k8sOrchestratorInstance        *K8sOrchestrator
+	onceFork8sOrchestratorInstance sync.Once
 )
 
+// FSSConfigMapInfo contains details about the FSS configmap(s) present in all flavors
+type FSSConfigMapInfo struct {
+	featureStates      map[string]string
+	configMapName      string
+	configMapNamespace string
+}
+
 // K8sOrchestrator defines set of properties specific to K8s
 type K8sOrchestrator struct {
-	featureStates   map[string]string
+	supervisorFSS   FSSConfigMapInfo
+	internalFSS     FSSConfigMapInfo
 	informerManager *k8s.InformerManager
+	clusterFlavor   cnstypes.CnsClusterFlavor
+}
+
+// K8sGuestInitParams lists the set of parameters required to run the init for K8sOrchestrator in Guest cluster
+type K8sGuestInitParams struct {
+	InternalFeatureStatesConfigInfo   config.FeatureStatesConfigInfo
+	SupervisorFeatureStatesConfigInfo config.FeatureStatesConfigInfo
+}
+
+// K8sSupervisorInitParams lists the set of parameters required to run the init for K8sOrchestrator in Supervisor cluster
+type K8sSupervisorInitParams struct {
+	SupervisorFeatureStatesConfigInfo config.FeatureStatesConfigInfo
+}
+
+// K8sVanillaInitParams lists the set of parameters required to run the init for K8sOrchestrator in Vanilla cluster
+type K8sVanillaInitParams struct {
+	InternalFeatureStatesConfigInfo config.FeatureStatesConfigInfo
 }
 
 // Newk8sOrchestrator instantiates K8sOrchestrator object and returns this object
-func Newk8sOrchestrator(ctx context.Context, featureStatesConfigMapInfo config.FeatureStatesConfigInfo) (*K8sOrchestrator, error) {
-	var coInstanceErr error
+// NOTE: As Newk8sOrchestrator is created in the init of the driver and syncer components,
+// raise an error only if it is of utmost importance
+func Newk8sOrchestrator(ctx context.Context, controllerClusterFlavor cnstypes.CnsClusterFlavor, params interface{}) (*K8sOrchestrator, error) {
+	var (
+		coInstanceErr error
+		k8sClient     clientset.Interface
+	)
 	onceFork8sOrchestratorInstance.Do(func() {
 		log := logger.GetLogger(ctx)
 		log.Info("Initializing k8sOrchestratorInstance")
 		k8sOrchestratorInstance = &K8sOrchestrator{}
-		k8sOrchestratorInstance.featureStates = make(map[string]string)
-		k8sClient, coInstanceErr := k8s.NewClient(ctx)
+		k8sOrchestratorInstance.clusterFlavor = controllerClusterFlavor
+
+		// Create a K8s client
+		k8sClient, coInstanceErr = k8s.NewClient(ctx)
 		if coInstanceErr != nil {
 			log.Errorf("Creating Kubernetes client failed. Err: %v", coInstanceErr)
```

Comment:
@SandeepPissay the new implementation supports the following:
1. k8sorchestrator will continue to be a singleton
2. It will support retries but the retry mechanism is not implemented. We can take it up in the future if required.
3. On receiving an error in creating any internmediate client objects, we will return `nil, error` instead of returning a half-baked object.
4. Not using `Once.Do` anymore

Can you confirm if this resolves your concern?